File: data-structures/Tree/BinarySearchTree/java/BinarySearchTree.java
/**
 * BinarySearchTree
 */
public class BinarySearchTree {

    public TreeNode root;

    public void add(int data){
        TreeNode node = new TreeNode(data);
        if(this.root == null){
            this.root = node;
        }else{
            this.insertNode(this.root,data);
        }
    }

    private void insertNode(TreeNode root,int data){
        TreeNode node = new TreeNode(data);
        if(data < root.data){
            if(root.left == null){
                root.left = node;
            }else{
                this.insertNode(root.left,data);
            }
        }else{
            if(root.right == null){
                root.right = node;
            }else{
                this.insertNode(root.right,data);
            }
        }
    }

    public boolean search(int data){
        return this.find(data) != null;
    }

    public TreeNode find(int data){
        return find(this.root,data);
    }

    private TreeNode find(TreeNode root,int data){
        if(root == null)return null;
        if(data == root.data){
            return root;
        }else if(data < root.data){
            return find(root.left,data);
        }else{
            return find(root.right,data);
        }
    }

    public TreeNode findParent(int data){
        return findParent(this.root,data);
    }

    private TreeNode findParent(TreeNode root,int data){
        if(root == null)return null;
        if(root.data == data) return null;
        if(data < root.data){
            if(root.left == null){
                return null;
            }else if(data == root.left.data){
                return root;
            }else{
                return findParent(root.left,data);
            }
        }else{
            if(root.right == null){
                return null;
            }else if(data == root.right.data){
                return root;
            }else{
                return findParent(root.right,data);
            }
        }
    }

    public boolean remove(int data){
        TreeNode nodeToRemoved = find(data);

        if(null == nodeToRemoved)return false;
        
        TreeNode parent = findParent(data);
        if (this.root.left == null && this.root.right == null){
            this.root = null;
        }else if(nodeToRemoved.left == null && nodeToRemoved.right == null){
            if(nodeToRemoved.data < parent.data){
                parent.left = null;
            }else{
                parent.right = null;
            }
        }else if(nodeToRemoved.left == null && nodeToRemoved.right != null){
            if(nodeToRemoved.data < parent.data){
                parent.left = nodeToRemoved.right;
            }else{
                parent.right = nodeToRemoved.right;
            }
        }else if(nodeToRemoved.left != null && nodeToRemoved.right == null){
            if(nodeToRemoved.data < parent.data){
                parent.left = nodeToRemoved.left;
            }else{
                parent.right = nodeToRemoved.left;
            }
        }else{
            
            TreeNode largestValue = nodeToRemoved.left;
            while (largestValue.right != null) {
                largestValue = largestValue.right;
            }
            this.remove(largestValue.data);
            nodeToRemoved.data = largestValue.data;
        }
        return true;
    }

    public void preOrder(){
        preOrder(this.root);
    }

    private void preOrder(TreeNode root){
        if(root != null){
            System.out.println(root.data);
            preOrder(root.left);
            preOrder(root.right);
        }
    }

    public void inOrder(){
        inOrder(this.root);
    }

    private void inOrder(TreeNode root){
        if(root != null){
            inOrder(root.left);
            System.out.println(root.data);
            inOrder(root.right);
        }
    }

    public void postOrder(){
        postOrder(this.root);
    }

    private void postOrder(TreeNode root){
        if(root != null){
            postOrder(root.left);
            postOrder(root.right);
            System.out.println(root.data);
        }
    }


    public static void main(String[] args){
        BinarySearchTree tree = new BinarySearchTree();
        tree.add(10);
        tree.add(2);
        tree.add(12);
        tree.add(4);
        tree.add(1);

        System.out.println(tree.root.data+" left: "+tree.root.left.data+" right: "+tree.root.right.data);
        tree.remove(10);
        tree.preOrder();
        
    }
}

File: data-structures/Tree/BinarySearchTree/go/BinarySearchTree.go
package main

import (
	"fmt"
)

type TreeNode struct {
	data  int
	left  *TreeNode
	right *TreeNode
}

type BinarySearchTree struct {
	root *TreeNode
}

func (tree *BinarySearchTree) add(data int) {
	if tree.root == nil {
		tree.root = &TreeNode{data: data, left: nil, right: nil}
	} else {
		add(tree.root, data)
	}
}

func add(root *TreeNode, data int) {
	if data < root.data {
		if root.left == nil {
			root.left = &TreeNode{data: data, left: nil, right: nil}
		} else {
			add(root.left, data)
		}
	} else {
		if root.right == nil {
			root.right = &TreeNode{data: data, left: nil, right: nil}
		} else {
			add(root.right, data)
		}
	}
}

func (tree *BinarySearchTree) find(data int) *TreeNode {
	return find(tree.root, data)
}

func find(root *TreeNode, data int) *TreeNode {
	if root == nil {
		return nil
	}
	if root.data == data {
		return root
	} else if data < root.data {
		return find(root.left, data)
	} else {
		return find(root.right, data)
	}

}

func (tree *BinarySearchTree) findParent(data int) *TreeNode {
	return findParent(tree.root, data)
}

func findParent(root *TreeNode, data int) *TreeNode {
	if root == nil {
		return nil
	}
	if data == root.data {
		return nil
	} else if data < root.data {
		if root.left == nil {
			return nil
		} else if data == root.left.data {
			return root
		} else {
			return findParent(root.left, data)
		}
	} else {
		if root.right == nil {
			return nil
		} else if data == root.right.data {
			return root
		} else {
			return findParent(root.right, data)
		}
	}
}

func (tree *BinarySearchTree) remove(data int) bool {
	nodeToRemoved := tree.find(data)
	if nodeToRemoved == nil {
		return false
	}
	parent := tree.findParent(data)
	if tree.root.left == nil && tree.root.right == nil {
		tree.root = nil
		return true
	} else if nodeToRemoved.left == nil && nodeToRemoved.right == nil {
		if nodeToRemoved.data < parent.data {
			parent.left = nil
		} else {
			parent.right = nil
		}
	} else if nodeToRemoved.left == nil && nodeToRemoved.right != nil {
		if nodeToRemoved.data < parent.data {
			parent.left = nodeToRemoved.right
		} else {
			parent.right = nodeToRemoved.right
		}
	} else if nodeToRemoved.left != nil && nodeToRemoved.right == nil {
		if nodeToRemoved.data < parent.data {
			parent.left = nodeToRemoved.left
		} else {
			parent.right = nodeToRemoved.left
		}
	} else {
		largestValue := nodeToRemoved.left
		for largestValue.right != nil {
			largestValue = largestValue.right
		}
		tree.remove(largestValue.data)
		nodeToRemoved.data = largestValue.data
	}
	return true
}

func (tree *BinarySearchTree) preOrder() {
	preOrder(tree.root)
}

func preOrder(root *TreeNode) {
	if root != nil {
		fmt.Println(root.data)
		preOrder(root.left)
		preOrder(root.right)
	}
}

func (tree *BinarySearchTree) inOrder() {
	inOrder(tree.root)
}

func inOrder(root *TreeNode) {
	if root != nil {
		inOrder(root.left)
		fmt.Println(root.data)
		inOrder(root.right)
	}
}

func (tree *BinarySearchTree) postOrder() {
	postOrder(tree.root)
}

func postOrder(root *TreeNode) {
	if root != nil {
		postOrder(root.left)
		postOrder(root.right)
		fmt.Println(root.data)
	}
}

func main() {
	tree := BinarySearchTree{}
	tree.add(10)
	tree.add(12)
	tree.add(2)
	tree.add(4)
	tree.add(1)

	//tree.preOrder()
	fmt.Println(tree.remove(4))
	tree.preOrder()
}
